content/post/2024-02-27-keyring/index.Rmd

```diff
@@ -0,0 +1,104 @@
+---
+slug: key-advantages-of-using-keyring
+title: "Key advantages of using the keyring package" 
+authors: 
+- Maëlle Salmon
+date: "2024-02-27" 
+tags: 
+- package development
+output: hugodown::hugo_document
+---
+
+```{r setup, include=FALSE}
+
+knitr::opts_chunk$set(fig.path = "", comment = "")
+# knitr hook to make images output use Hugo options
+knitr::knit_hooks$set(
+  plot = function(x, options) {
+    hugoopts <- options$hugoopts
+    paste0(
+      "{{<figure src=",
+      '"', x, '" ',
+      if (!is.null(hugoopts)) {
+        glue::glue_collapse(
+          glue::glue('{names(hugoopts)}="{hugoopts}"'),
+          sep = " "
+        )
+      },
+      ">}}\n"
+    )
+  }
+)
+
+```
+
+```{r, include = FALSE, cache = FALSE}
+knitr::opts_chunk$set(
+  collapse = TRUE,
+  comment = "#>",
+  out.width = "100%",
+  asciicast_theme = "pkgdown"
+)
+```
+
+Does your package need the user to provide secrets, like API tokens, to work?
+Have you considered telling your package users about the keyring package?
+
+The [keyring](https://keyring.r-lib.org/) package maintained by Gábor Csárdi is a package that accesses the **system credential store** from R: each operating system has a special place for storing secrets securely, that keyring knows how to interact with.
+The credential store can hold several **keyrings**, each **keyring** can be protected by a specific password and can hold several **keys** which are the **secrets**.
+
+Why would one use keyring?
+
+## A simple example
+
+Let's exemplify the usage of keyring with the example of `mypkg::my_fun()` that needs an API key to work and expects it to be available as `Sys.getenv("MYSECRET")`.
+This is already a more secure approach that having the API key as an argument, which could lead to users writing their secret in their script, and from there maybe inadvertently sharing the script thus the secret.
+
+### Without using keyring
+
+1.  The user set an environment variable `MYSECRET="super-secret"`
+    -   in a project-specific [`.Renviron`](https://rstats.wtf/r-startup#renviron), hopefully [.gitignored](https://usethis.r-lib.org/reference/use_git_ignore.html);
+    -   or in the user `.Renviron`;
+    -   or types `Sys.setenv(MYSECRET="super-secret")` in the console, hopefully not saving, not sharing `.Rhistory`.
+2.  The package retrieves the secret with `Sys.getenv("MYSECRET")`.
+
+### With keyring
+
+1.  The user stores the secret, once and for all per computer, using `keyring::key_set("MYSECRET")`, typing interactively so nothing is recorded in `.Rhistory`.
+2.  The user sets an environment variable with `Sys.setenv(MYSECRET = keyring::key_get("MYSECRET"))` in a script for instance.
+3.  The package retrieves the secret with `Sys.getenv("MYSECRET")`.
```

Comment:
Alternatively, packages may support keyring directy, and try to retrieve secrets directly. I wonder if there is any existing package that does that. gh used to, before switching to the git credential store.

Reply:
actually yes, for instance https://github.com/bluegreen-labs/ecmwfr/blob/534e98831c0af9859bcc70a48258d804f2d62108/R/wf_get_key.R#L23 that I'll add as an example